gcm/benchmarks/benchmarking_test.go

```diff
@@ -0,0 +1,300 @@
+	received      int           // received messages
+
+	service  *server.Service
+	receiveC chan bool
+	doneC    chan struct{}
+
+	wg    sync.WaitGroup
+	start time.Time
+	end   time.Time
+}
+
+func (params *benchParams) String() string {
+	return fmt.Sprintf(`
+		Throughput %.2f messages/second using:
+			%d workers
+			%d gcm subscriptions
+			%s gcm response timeout
+			%d clients
+	`, params.mps(), params.workers, params.subscriptions, params.timeout, params.clients)
+}
+
+func (params *benchParams) expectedMessagesCount() int {
+	return params.N * params.clients * params.subscriptions
+}
+
+func (params *benchParams) send(c client.Client) error {
+	err := params.sender(c)
+	if err != nil {
+		return err
+	}
+	params.sent++
+	return nil
+}
+
+func (params *benchParams) receiveLoop() {
+	for i := 0; i <= params.workers; i++ {
+		go func() {
+			for {
+				select {
+				case <-params.receiveC:
+					params.received++
+					logger.WithField("received", params.received).Info("Received gcm call")
+					params.wg.Done()
+				case <-params.doneC:
+					return
+				}
+			}
+		}()
+	}
+}
+
+// start the service
+func (params *benchParams) setUp() {
+	params.doneC = make(chan struct{})
+	params.receiveC = make(chan bool)
+
+	a := assert.New(params)
+
+	dir, errTempDir := ioutil.TempDir("", "guble_benchmarking_gcm_test")
+	defer func() {
+		errRemove := os.RemoveAll(dir)
+		if errRemove != nil {
+			logger.WithFields(log.Fields{"module": "testing", "err": errRemove}).Error("Could not remove directory")
+		}
+	}()
+	a.NoError(errTempDir)
+
+	*config.Listen = "localhost:0"
+	*config.KVBackend = "memory"
+	*config.MSBackend = "file"
+	*config.StoragePath = dir
+	*config.GCM.Enabled = true
+	*config.GCM.APIKey = "WILL BE OVERWRITTEN"
+	*config.GCM.Workers = params.workers
+
+	params.service = gubled.StartService()
+
+	gcmConnector, ok := params.service.Modules()[4].(*gcm.Connector)
+	a.True(ok, "Modules[4] should be of type GCMConnector")
+
+	gcmConnector.Sender = testutil.CreateGcmSender(
+		testutil.CreateRoundTripperWithCountAndTimeout(http.StatusOK, testutil.SuccessGCMResponse, params.receiveC, params.timeout))
+
+	urlFormat := fmt.Sprintf("http://%s/gcm/%%d/gcmId%%d/subscribe/%%s", params.service.WebServer().GetAddr())
+	for i := 1; i <= params.subscriptions; i++ {
+		// create GCM subscription with topic: gcmTopic
+		response, errPost := http.Post(
+			fmt.Sprintf(urlFormat, i, i, strings.TrimPrefix(gcmTopic, "/")),
+			"text/plain",
+			bytes.NewBufferString(""),
+		)
+		a.NoError(errPost)
+		a.Equal(response.StatusCode, 200)
+
+		body, errReadAll := ioutil.ReadAll(response.Body)
+		a.NoError(errReadAll)
+		a.Equal("registered: /topic\n", string(body))
+	}
+}
+
+func (params *benchParams) tearDown() {
+	assert.NoError(params, params.service.Stop())
+	params.service = nil
+}
+
+func (params *benchParams) createClients() []client.Client {
+	wsURL := "ws://" + params.service.WebServer().GetAddr() + "/stream/user/"
```

Comment:
maybe put this on the top of the file alongside with other variables likes totpic